### ISSUE TITLE
Document a way to determined java install is OpenJ9

### ISSUE THREAD

**Issue by headius** [eclipse/openj9#6760]
For JRuby we would like to be able to have JVM-specific flags as part of our default launcher. These flags are used to improve startup or tweak other settings to tune the target JVM for performance.

We would like to be able to do this without first launching the discovered `java` command. We want a reliable way to determine, from simple files in the OpenJ9 JAVA_HOME, that we will eventually launch using OpenJ9.

As an example, we would like to pass `-Xquickstart -Xshareclasses` for our faster-starting `--dev` mode, but we can only add those options if we will run on OpenJ9. Similarly, we use flags for Hotspot that either have no meaning or have different equivalents on OpenJ9, such as `-XX:TieredStopAtLevel=1`.

Ideally whatever this mechanism is, it should work on current releases of OpenJ9 back to Java 8, or be backported. If we can't detect OpenJ9 we'll just assume it's Hotspot, which mostly works.

Does any such mechanism exist?

**Comment by mstoodle** [comment]
@headius You can solve part of the problem via environment variables without having to identify the JVM.

There is an environment variable `OPENJ9_JAVA_OPTIONS` that can be set to OpenJ9-specific (or any) options that will be processed in addition to the command-line options (with rightmost command-line option overriding any conflicts).

e.g. (i didn't have the most recent JDK8 build handy, so forgive older version info)

```
$ OPENJ9_JAVA_OPTIONS="-Xshareclasses:verbose" $OPENJ9JDK/bin/java -version
[-Xshareclasses persistent cache enabled]
[-Xshareclasses verbose output enabled]
JVMSHRC237I Opened shared classes persistent cache sharedcc_mstoodle
JVMSHRC246I Attached shared classes persistent cache sharedcc_mstoodle
JVMSHRC765I Memory page protection on runtime data, string read-write data and partially filled pages is successfully enabled
openjdk version "1.8.0_212"
OpenJDK Runtime Environment (build 1.8.0_212-b03)
Eclipse OpenJ9 VM (build openj9-0.14.0, JRE 1.8.0 Mac OS X amd64-64-Bit Compressed References 20190417_217 (JIT enabled, AOT enabled)
OpenJ9   - bad1d4d06
OMR      - 4a4278e6
JCL      - 5590c4f818 based on jdk8u212-b03)
JVMSHRC168I Total shared class bytes read=2183280. Total bytes stored=1904
JVMSHRC818I Total unstored bytes due to the setting of shared cache soft max is 0. Unstored AOT bytes due to the setting of -Xscmaxaot is 0. Unstored JIT bytes due to the setting of -Xscmaxjitdata is 0.
```
As you can see, the J9-specific `-Xshareclasses` was processed from the environment variable. Hotspot isn't bothered by the environment variable:

```
$ OPENJ9_JAVA_OPTIONS="-Xshareclasses:verbose" $HOTSPOT_JDK/bin/java -version
openjdk version "1.8.0_212"
OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_212-b03)
OpenJDK 64-Bit Server VM (AdoptOpenJDK)(build 25.212-b03, mixed mode)
```

Unfortunately, I don't know (maybe someone else does) a way to provide options to Hotspot that won't also be processed by OpenJ9. You can *sort of* handle this by adding `-XX:+IgnoreUnrecognizedVMOptions` to the OpenJ9 options, but that isn't perfect, particularly if you ever want to set an option one way for OpenJ9 but another way for Hotspot.

e.g.

```
$ OPENJ9_JAVA_OPTIONS="-XX:+IgnoreUnrecognizedVMOptions" $OPENJ9JDK/bin/java -XX:TieredStopAtLevel=1 -version
openjdk version "1.8.0_212"
OpenJDK Runtime Environment (build 1.8.0_212-b03)
Eclipse OpenJ9 VM (build openj9-0.14.0, JRE 1.8.0 Mac OS X amd64-64-Bit Compressed References 20190417_217 (JIT enabled, AOT enabled)
OpenJ9   - bad1d4d06
OMR      - 4a4278e6
JCL      - 5590c4f818 based on jdk8u212-b03)
```

**Since it's the same OpenJ9 built into every JDK release, this approach will work on every release from JDK8 and forward (and even on IBM SDK for Java 8, fwtw :) ).**

**Comment by smlambert** [comment]
You could query system property, java.vm.name

example in Java:
https://github.com/eclipse/openj9/blob/master/test/TestConfig/src/org/openj9/envInfo/JavaInfo.java#L101

example from command-line:
java -XshowSettings:properties -version  (and find the java.vm.name property in the output)

**Comment by pshipton** [comment]
FYI that OpenJ9 ignores unrecognized options starting with `-XX:` by default, controlled by https://www.eclipse.org/openj9/docs/xxignoreunrecognizedxxcolonoptions/

Without running java, there are a number of files in the build which are unique to OpenJ9. `openj9-notices.html` is one such, found in the root directory for jdk8 and the lib directory for jdk11 and later. `lib\options.default` is another file unique to OpenJ9 builds, which needs to exist or there is a warning when the JVM starts. There are other files, but they could be excluded when using jlink with jdk11+.

The `release` file can be used, but it only contains enough information from jdk11 on.
`SOURCE="OpenJDK:9fad1c64af OpenJ9:412258978 OMR:4a4278e6"`

**Comment by mstoodle** [comment]
@headius any comment on the above suggestions? Are any of them workable for what you're trying to do?

**Comment by headius** [comment]
Hey folks, thanks for the suggestions!

@mstoodle The environment variable is a great workaround for now! It will mean setting additional noise into the environment even when not on OpenJ9, but at least I can update our launcher to properly enable quickstart+shareclasses in dev mode!

It's unfortunate there's not a similar environment for Hotspot-only options, but [the ones I'm passing](https://github.com/jruby/jruby/pull/5826/commits/668e69c7496b6cefe1b39f0ddcd1f8614f1e0e20) appear to be either valid (-Xverify:none) or ignored (-XX:Tiered*)

@smlambert If we could accept the hit of launching the JVM that would work fine, but as it is that would add anywhere from 0.1 to 0.3s to our already-slow startup. I need a mechanism that works without launching the JVM (or that gets me the information I need in under, say, 0.05s).

@pshipton I did notice the `release` file but I was looking at Java 8, which only shows Java compatibility level (not VM information). The other files might be usable for this purpose, but that makes me a little nervous; they just have to move or change name in a future release and they'll stop being detected. 😬

I will look at using the environment variable for now in jruby/jruby#5826. I assume OpenJ9 from Java 9 also supports `@arguments` files, yes?

**Comment by DanHeidinga** [comment]
> I assume OpenJ9 from Java 9 also supports @arguments files, yes?

Yes, OpenJ9 supports the @arguments files.  We also support options files in JDK8 using `-Xoptionsfile` (https://www.eclipse.org/openj9/docs/xoptionsfile/#xoptionsfile)

**Comment by headius** [comment]
In order for me to use arguments files the J9 way I'd need a way to determine I'm on J9 first, which is still the crux of this issue. However I have introduced the environment variable into our laucher's "dev" mode here:

https://github.com/jruby/jruby/pull/5826/commits/01a6e07db6d1b88920b612d04ad749f6590ac675
